addons/account/models/account_move.py

```diff
@@ -990,14 +990,16 @@ def _recompute_dynamic_lines(self, recompute_all_taxes=False, recompute_tax_base
                                     or not depending of the field 'recompute_tax_line' in lines.
         '''
         for invoice in self:
+            recompute_invoice_tax_lines = recompute_all_taxes
             # Dispatch lines and pre-compute some aggregated values like taxes.
-            for line in invoice.line_ids:
-                if line.recompute_tax_line:
-                    recompute_all_taxes = True
-                    line.recompute_tax_line = False
+            if any(line.recompute_tax_line for line in invoice.line_ids) or\
+                invoice.mapped('line_ids.tax_ids').flatten_taxes_hierarchy()._origin > invoice.mapped('line_ids.tax_line_id')._origin:
+                recompute_invoice_tax_lines = True
+                invoice.line_ids.recompute_tax_line = False
+
 
             # Compute taxes.
-            if recompute_all_taxes:
+            if recompute_invoice_tax_lines:
```

Comment:
* `mapped` is useless for relational fields and just makes the lines longer and harder to read
* operators should be at the start of lines on multiple lines statements https://peps.python.org/pep-0008/#should-a-line-break-before-or-after-a-binary-operator
* parenthesis is preferred to backslashes to continue lines (more consistent indentation) https://peps.python.org/pep-0008/#maximum-line-length
In this case, you have a statement that is indented on the same level as the condition above without any clear separation
* just personal opinion: I think that adding a new variables (`recompute_invoice_tax_lines`) adds confusion

```suggestion
            if (
                recompute_all_taxes
                or any(line.recompute_tax_line for line in invoice.line_ids)
                or invoice.line_ids.tax_ids.flatten_taxes_hierarchy()._origin > invoice.line_ids.tax_line_id._origin
            ):
                invoice.line_ids.recompute_tax_line = False
```


I'm not sure why you need the `_origin` though. What happens if you remove it?

Reply:
> for `recompute_invoice_tax_lines`: The reason for adding it is that the previous code set `recompute_tax_lines` to True when an invoice matches the condition, even if there are 10^6 invoices and only the first one needs to be recomputed. If your concern is about the name, do you have any recommendation,? what about `recompute_current_invoice_tax_lines` or `invoice_recompute_tax_lines`?

Note that I'm not changing the value of `recompute_all_taxes` anymore. The diff I proposed is all you need in this PR (except tests)

Reply:
You can remove all the lines that contain `recompute_invoice_tax_lines` in https://github.com/odoo/odoo/commit/f1e6d1e3cc9d3e811fb42ea2f24cf41446f33154